### PR TITLE
Create IL Rewriting Basics.md

### PR DESCRIPTION
**Pull request by davmason**
@williamXieMSFT asked me to create a short article on IL rewriting and point to CLR Instrumentation Engine so folks who are interested in profiling know about CLRIE.

Is this along the lines of what you expected? I can flesh out areas more if necessary (although adding full instructions on how to rewrite IL would not be feasible in the short term).

### CODE REVIEW

File: Documentation/Profiling/IL Rewriting Basics.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  - Inspecting managed objects
+  - Inspecting function arguments/return values
+- Injecting method hooks at the start/end of the method that call in to another managed library
+
+There are two ways to rewrite IL
+
+1. At Module load time with `ICorProfilerInfo::SetILFunctionBody`
+    This approach has the benefit that it is 'set it and forget it'. You can replace the IL at module load, and the runtime will treat this new IL as if the module contained that IL - you don't have to worry about any of the quirks of ReJIT. The downside is that is is unrevertable - once it is set, you cannot change your mind.
+
+2. At any point during the process lifetime with 'ICorProfilerInfo4::RequestReJIT' or 'ICorProfilerInfo10::RequestReJITWithInliners'.
```
Comment (mattwarren) [inline Documentation/Profiling/IL Rewriting Basics.md]: If you want to be consistent with the other places in the doc, these need to be surrounded with backticks, i.e:

\`ICorProfilerInfo4::RequestReJIT\` or \`ICorProfilerInfo10::RequestReJITWithInliners\`

File: Documentation/Profiling/IL Rewriting Basics.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+   This approach means that you can modify functions in response to changing conditions, and you can revert the modified code if you decide you are done with it. See the other entries about ReJIT in this folder for more information.
+
+## How to rewrite IL
+Hopefully this section will be fleshed out in the future. Right now we have some documentation in the archives at [Creating an IL-rewriting profiler](davbr-blog-archive/Creating an IL-rewriting profiler.md), but there is no start to finish tutorial on IL rewriting.
```
Comment (mattwarren) [inline Documentation/Profiling/IL Rewriting Basics.md]: GitHub isn't rendering/detecting the relative link to 'davbr-blog-archive/Creating an IL-rewriting profiler.md' properly? 

Maybe there's a way to get it working, but I suspect just changing it to the full URL might be easier/quicker!

Comment (gfoidl) [reply Documentation/Profiling/IL Rewriting Basics.md]: It's not github, as written it's a invalid link destination according to the [CommonMark-spec](https://spec.commonmark.org/0.29/#link-destination) due the whitespaces in the url.

So the need to be "escaped".
```suggestion
Hopefully this section will be fleshed out in the future. Right now we have some documentation in the archives at [Creating an IL-rewriting profiler](<./davbr-blog-archive/Creating an IL-rewriting profiler.md>), but there is no start to finish tutorial on IL rewriting.
```

Comment (WilliamXieMSFT) [reply Documentation/Profiling/IL Rewriting Basics.md]: FWIW `%20` for spaces also works

---
In reply to: [329039442](https://github.com/dotnet/coreclr/pull/26918#discussion_r329039442) [](ancestors = 329039442)

**Comment by benaadams** [comment]
Corefx failure regex timeout
```
System.Text.RegularExpressions.Tests.RegexGroupTests.GroupsTurkish
Timed out at 9/27/2019 9:14:58 AM after 60000ms waiting for remote process.
```

File: Documentation/Profiling/IL Rewriting Basics.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
+# IL Rewriting Basics
+
+## Intro
+One of the common use cases of the `ICorProfiler*` interfaces is to perform IL rewriting. Some possible reasons a profiler would want to rewrite IL:
+- Inspecting interesting process state
+  - Capturing exception state
+  - Inspecting managed objects
+  - Inspecting function arguments/return values
+- Injecting method hooks at the start/end of the method that call in to another managed library
```
Comment (MarcoRossignoli) [inline Documentation/Profiling/IL Rewriting Basics.md]: Could be used for injecting method at specific point, for instance for "code coverage instrumentation"?

File: Documentation/Profiling/IL Rewriting Basics.md (hunk 1 above)
Comment (davmason) [reply Documentation/Profiling/IL Rewriting Basics.md]: Thanks!

**Comment by davmason** [comment]
@noahfalk  I didn't have any specific plans for discoverability. @WilliamXieMSFT asked that CLRIE be mentioned in the profiling docs and this covers that request. I don't believe there is a need to advertise, I think the idea is just that anyone who is starting with IL rewriting will look at the profiler docs and see this.

File: Documentation/Profiling/IL Rewriting Basics.md (hunk 3 above)
Comment (MarcoRossignoli) [reply Documentation/Profiling/IL Rewriting Basics.md]: > ELT hooks 

From guide seem that ELT is too coarse(enter/exit func) for collect line/branch...but thank's for infos!
https://github.com/dotnet/coreclr/blob/master/Documentation/Profiling/davbr-blog-archive/ELT%20Hooks%20-%20The%20Basics.md

**Comment by noahfalk** [comment]
> anyone who is starting with IL rewriting will look at the profiler docs and see this.

My anecdotal evidence is that when docs only exist in our repo with nothing else guiding people to them (good web search placement, MSDN links) then most potential users do not find them. That said I am perfectly happy to keep this as-is and be reactive. If you see indications that the information isn't reaching people who would have benefitted from it you can always cast a wider net in the future.

**Pull request merged**